File: tasks/money-rollup/javascript/src/rollup.js
"use strict";

function rollup(items) {
  const totals = new Map();
  for (const [account, minor] of items) {
    totals.set(account, (totals.get(account) ?? 0n) + minor);
  }
  return [...totals.entries()].sort(([left], [right]) =>
    left < right ? -1 : left > right ? 1 : 0,
  );
}

module.exports = { rollup };

File: tasks/money-rollup/go/src/rollup.go
package main

import (
	"math/big"
	"slices"
)

type item struct {
	account string
	minor   *big.Int
}

func rollup(items []item) []item {
	totals := map[string]*big.Int{}
	for _, entry := range items {
		total, present := totals[entry.account]
		if !present {
			total = new(big.Int)
			totals[entry.account] = total
		}
		total.Add(total, entry.minor)
	}
	accounts := make([]string, 0, len(totals))
	for name := range totals {
		accounts = append(accounts, name)
	}
	slices.Sort(accounts)
	rows := make([]item, 0, len(accounts))
	for _, name := range accounts {
		rows = append(rows, item{account: name, minor: totals[name]})
	}
	return rows
}
